### ISSUE TITLE
WISH: Add support for latest Pindel

### ISSUE THREAD

**Issue by shuntsman-ucsf** [UCSF-Costello-Lab/LG3_Pipeline#122]
We would like to request Pindel to be updated to the latest version 0.2.5b8 if possible.  In addition to improved features, there are also issues with the old version that cause samples to fail. I think Stephanie also prefers the new version.

https://github.com/genome/pindel

**Comment by ivan108** [comment]
OK, I will work on it...

**Comment by HenrikBengtsson** [comment]
... and I'll try to install the latest version of Pindel as a module (as soon as the sluggishness of the cluster goes away)

**Comment by HenrikBengtsson** [comment]
Pindel 0.2.5b8 installed (https://github.com/UCSF-TI/TIPCC/issues/249).  @ivan108, in the scripts this specific version can be loaded as:
```sh
module load CBC pindel/0.2.5b8
```

**Comment by ivan108** [comment]
Here is my first experience of using "new" pindel in LG3 pipline:
Step 1. Generating config file using scripts/pindel_setup.sh == Success
Step 2. Running pindel == Success. Note: It seems new pindel is much more sensitive, I got >23k hits instead of ~3k!
Step 3. Converting pindel output into VCF using pindel2vcf utility provided by pindel == Success
Step 4. Converting VCF into custom tab-delim format using custom vcf2tdt.py script == Failed, see error:
```
  File "/home/jocostello/shared/LG3_Pipeline/scripts/vcf2tdt.py", line 68, in vcf2tdt
    format_dict[n] = l[format].split(':').index(n)
ValueError: list.index(x): x not in list
```

So, I compared VCF file produced by "new" pindel to VCF I got for older pindel. Format is almost identical, except that in new VCF AD field shows Allele Depth for both alleles (with comma separation), while in old one just one. I guess that breaks parsing by Python script...

Possible solutions:
1. figure out Python script and fix it
2. try existing VCF to tsv converters to achieve the same result (preferable?)

**Comment by shuntsman-ucsf** [comment]
Can you post the paths to one each of the old and new vcfs?  Changing the python script might be doable.

________________________________

**Comment by ivan108** [comment]
Here they are
Old 
`/costellolab/data1/jocostello/LG3/pindel/Patient157t10_pindel.orig/Patient157t10.pindel.vcf`
New 
`/costellolab/data1/jocostello/LG3/pindel/Patient157t10_pindel/Patient157t10.pindel.vcf`

BTW, I compared help files for the old and new pindel trying to understand the reason for higher sensitivity of the new pindel. It turns out that in the old pindel the parameter -M/--minimum_support_for_event is defaulted to 3, while in the new one to 1(!). That should explain at least some difference in sensitivity.. Another thing, in the new pindel there is a parameter -E/--sensitivity (default 0.95) which can be adjusted. There is no such a parameter in the old pindel...

**Comment by shuntsman-ucsf** [comment]
The new vcf also has additional FORMAT fields listed in the header for 'PL' and 'RD', although they are not found in the data (at least in this 1 sample).  The error occurs when the script is iterating through the data expecting the additional fields but finds less and goes out of range.  I don't know PYTHON (more of a PERL guy) but can try to make a fix.

**Comment by ivan108** [comment]
You are right, I noticed those extra FORMAT fields too, but didn't relate them to the problem. The easiest fix is just to delete those fields from the header, since they are not used in the body anyway. I did just that and it worked! I was able to make a next step, generate "tdt" file. 
However, next filtering step has failed. This time it looks that comma-separated AD values are breaking parsing in scripts/pindel_filter.py script:

  ```
File "/home/jocostello/shared/LG3_Pipeline/scripts/pindel_filter.py", line 197, in <module>
    pindel_filter(args[0], opts)
  File "/home/jocostello/shared/LG3_Pipeline/scripts/pindel_filter.py", line 45, in pindel_filter
    if int(l[norm_alt]) > opts.normal: continue
ValueError: invalid literal for int() with base 10: '16,3'
```

**Comment by shuntsman-ucsf** [comment]
Then it seems the meaning and format of AD has changed between pindel or vcf versions.  I'm not sure if the old pindel broke format or if it is a vcf version issue.  Recent files I have used usually have the comma separated values representing allele depth for each allele.  DP would be a single value for both, but this might be the value for just the ALT.  

Either way, the scripts would have to be changed in order to accomodate it, but not until we are sure we are referring to the correct values in each file and know what needs to be changed or parsed from it.

**Comment by ivan108** [comment]
I analyzed the post-pindel code, including problematic pindel_filter.py, and it is clear that the single AD  value in the "old" VCF is indeed a depth of ALT allele. I think that is against VCF4.0 specs, but I couldn't confirm it. The depth of ALT allele for Normal and Tumor samples is used to filter pindel raw hits. The default settings to drop the hit is : AD[normal] > 0 and AD[tumor]  < 6. There are other filters, but only this one is based on AD values. As for the "new" VCF, I believe the first AD value is for REF allele, and the second - for ALT. Assuming this is true I modified pindel_filter.py script to extract second AD value and use it for filtering. All pindel scripts are still backward compatible. One can choose which pindel version to use by editing PINDEL_NEW parameter (possible values: true or false) in lg3.conf file.

**Comment by shuntsman-ucsf** [comment]
Great work. Thanks.  But how do I access the new version as end user?  I don't think I have access to the conf file.  How would I pass that parameter, or does it require a new release for me to use?

**Comment by ivan108** [comment]
Yes, you need to grab a new (current) release. It also has an improved Mutect2 implementation and an option to use a new bam preparation pipeline based on bwa mem as per current Best Practice guidelines...

**Comment by HenrikBengtsson** [comment]
Notes to myself but a question too: So, the `lg3.conf` file needs to be updated with `PINDEL_NEW=true`.  The question is: How can user who does not install lg3 locally set this?  Where should this file be located?

**Comment by ivan108** [comment]
Yes, the idea is that `${LG3_HOME}/lg3.conf` could be edited by user to fine tune the pipeline...

**Comment by HenrikBengtsson** [comment]
Got it. Then I think we should provide a way to have a user/folder specific setup too. Think /etc/bashrc and ~/.bashrc.

**Comment by ivan108** [comment]
Great idea! I didn't think about it as I am always installing lg3 locally. Do you have any suggestions on how to implement this? I guess user lg3.conf should overwrite the global one...

**Comment by HenrikBengtsson** [comment]
Yes, first sourcing `$LG3_HOME/lg3.conf` and then `./lg3.conf` if it exists.  I just pushed a first prototype of this. See commit d365ec4.

**Comment by ivan108** [comment]
Hmm, if I install lg3 locally, then `$LG3_HOME/lg3.conf` and  ./lg3.conf will point to the same file?

**Comment by HenrikBengtsson** [comment]
I did actually consider that case in my first prototype (commit 49afc628) where it was used to protected against an infinite loop.  Here it'll just source the same file twice - annoying but shouldn't do damage.  Either way, we can add the same code to "skip" in the `source_lg_conf()` function.

**Comment by HenrikBengtsson** [comment]
I've updated `source_lg_conf()` so the same file won't be sourced twice, cf. commit 07a8e81.

@ivan108, I've also undone your merge (sorry for not being clear).  ... and please stop committing as @josephfcostello - update your `git config` author information.

**Comment by shuntsman-ucsf** [comment]
Am I able to use the latest release via "module load" and my own lg3.conf file yet, or is it still being ironed out?

**Comment by HenrikBengtsson** [comment]
Ivan, since you're the one who had actuallt used lg3.conf and new Pindel, can you verify that the current develop branch works for Scott?

When confirmed, I can make a quick release that Scott can use.

**Comment by ivan108** [comment]
OK, I am re-testing pindel on develop branch...

**Comment by ivan108** [comment]
I have re-run old and new pindel on my local clone of develop branch, seems to be working fine...

**Comment by HenrikBengtsson** [comment]
Thanks, did you run it where LG3_HOME != PWD?

**Comment by HenrikBengtsson** [comment]
To clarify my question, did you run the way that Scott will do with a local lg3.conf?

**Comment by ivan108** [comment]
Now I tried to launch it from ./runs_demo and it fails:
`./_run_Pindel: line 28: LG3_OUTPUT_ROOT: parameter null or not set`
which makes sense because on line 23 readlink returns nothing...

**Comment by HenrikBengtsson** [comment]
But I don't think that's how Scott/README runs the pipeline

**Comment by ivan108** [comment]
I also tried a "module way":

mkdir lg3-demo
cd lg3-demo
lg3 --version
2019-03-23 (commit 749df93)
mkdir output
export PATIENT=Patient157t10
lg3 test setup
*** Setup
[OK] LG3_HOME: /home/shared/cbc/software_cbc/LG3_Pipeline-2019-03-23
/home/shared/cbc/software_cbc/LG3_Pipeline-2019-03-23/bin/lg3-test: line 244: R_LIBS_USER=~/R/%p-library/%v /opt/R/R-latest/bin/Rscript: No such file or directory
[... other tests are OK]]

**Comment by HenrikBengtsson** [comment]
I moved the issue on '... /opt/R/R-latest/bin/Rscript: No such file or directory' to Issue #125, because I'm 99.999% sure it's irrelevant to the problem of running/validating/testing the "new" Pindel.

**Comment by HenrikBengtsson** [comment]
@ivan108 [wrote](https://github.com/UCSF-Costello-Lab/LG3_Pipeline/issues/125#issuecomment-484654141):

> I continue to run test pipeline, currently _run_Recal step. When it is done, I will create empty ./lg3.conf file with single line:
> PINDEL_NEW=true
> and see what happens when I run _run_Pindel ...

PS. For our future selves, let's try to keep the comments where they belong.  The topic on testing `PINDEL_NEW` should be kept in this thread.

**Comment by ivan108** [comment]
I finished Pindel test run on lg3 module release 2019-03-23, and it did ignore `PINDEL_NEW=true` from the local ./lg3.conf file, which makes sense, since that version doesn't have yet support for that feature...
Can module mechanism be used to load current develop version instead of release?

**Comment by HenrikBengtsson** [comment]
There's no `lg3/develop` module on TIPCC - I thought about it in the past but it wouldn't make sense since it would keep being updated and therefore would have been unstable to use.   The README has a Section 'Installation notes' at the end - it's very easy to "install" locally.  To get the develop version, just use `git checkout develop`.

**Comment by HenrikBengtsson** [comment]
FYI, I just dropped (commit e7700240) _local_ `PINDEL_NEW` and that if-else-statement - more clear to just comment/comment-out.  Also, setting `PINDEL_NEW=true` in `./lg3.conf` would not do anything.

**Comment by ivan108** [comment]
Hmm, I am confused now. I assume Scott is running the "module" installation which requires additional local ./lg3.conf file to control version of Pindel and possibly other things in the future, and this is what we are trying to test?? 
I myself use "local" installation for develop branch all the time...

**Comment by ivan108** [comment]
I see your point with `PINDEL_NEW=true` with "module installation" ...

**Comment by HenrikBengtsson** [comment]
I just said that the develop branch is not available via a TIPCC CBI module.  In order to release a version that Scott can use, we need to validate.  In order to test/validate the develop branch, it needs to "installed".

> I myself use "local" installation for develop branch all the time...

I suspect you mean you check out the git repos and run within that folder.   In order to test it properly, you need to emulate how it is to run it as another user.  That is, you must not run it inside LG3_HOME.  The installation instructions show how to check out lg3 somewhere (e.g. /path/to/) and the set LG3_HOME accordingly.  (I've updated the README to also set `PATH=$LG3_HOME/bin:PATH` so that `lg3` is found).  That's all it takes to run it from anywhere.   (The modules are basically just setting `LG3_HOME` and `PATH` like that, see output of `module show lg3`).

**Comment by ivan108** [comment]
Thanks for your patience :)
I created emulated develop setup and running initial steps ...

**Comment by HenrikBengtsson** [comment]
I also ran the checks.  This discovered the blocking bug in Issue #129 that breaks the existing pipeline (independently on the "old" or "new" Pindel).  That issue needs to be resolved before continuing on this one.

**Comment by HenrikBengtsson** [comment]
UPDATE 2019-07-22: I've made updates to the LG3 Pipeline code that should fix Bug #129 such that it's now possible for the end-user to run with a newer version of Pindel.  I'm currently running the full set of tests (with the default as well as the user-overridden Pindel settings).  Hopefully, this will solve it.  If it does, I'll try release a new version of LG3 within 24 hours.

/cc @shuntsman-ucsf @eladziv

**Comment by HenrikBengtsson** [comment]
I've not verified that the new version works.  I've just released LG3 Pipeline 2019-07-22 (and made it available on TIPCC).  The version of Pindel can now be controlled via a `./lg3.conf` file in the project folder. To use the latest Pindel version, let it contain:
```sh
## Project-specific LG3 configuration

## "New" version of Pindel, cf. https://github.com/UCSF-Costello-Lab/LG3_Pipeline/issues/122
PINDEL_HOME=/home/shared/cbc/software_cbc/pindel-0.2.5b8
PINDEL=${PINDEL_HOME}/pindel
PINDEL2VCF=${PINDEL_HOME}/pindel2vcf
```
This will override the default values set in [`${LG3_HOME}/lg3.conf`](https://github.com/UCSF-Costello-Lab/LG3_Pipeline/blob/c371bc0b83fb17c160ceb4d1cd671a8c2c5114ce/lg3.conf#L37-L39).

FYI, you can also set things such a `PATIENT` etc. in the `./lg3.conf` file.
